### PR TITLE
TickWheelExecutor is a resource, not an allocation

### PR DESCRIPTION
**Pull request by rossabaker**
Composing two `F[A, F[Unit]]`'s together defeats the purpose of the `Resource` AST: any intermediate cancellations can leak a resource.  The only safe way to combine resources is through the resource interpreter.  This makes the `BlazeClientBuilder.resource` the primitive constructor, and expresses `allocate` in terms of it.

### CODE REVIEW

File: core/src/main/scala/org/http4s/internal/package.scala [hunk 1]
```diff
@@ -117,4 +119,47 @@ package object internal {
           }
       }
     }
+
+  /* Derived from https://github.com/typelevel/cats-effect/blob/c62c6c76b3066c500fca2f9e0897605bc12eb2a0/core/shared/src/main/scala/cats/effect/Resource.scala */
+  private[http4s] def interpretResource[F[_], A](resource: Resource[F, A])(
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/internal/package.scala]: This is a copy of `Client.toHttpApp`, which should be expressed in terms of this.  But let's agree on whether it should return `ExitCase[Throwable] => F[Unit]` or just `F[Unit]` first.

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClientBuilder.scala [hunk 2]
```diff
@@ -128,33 +128,28 @@ sealed abstract class BlazeClientBuilder[F[_]] private (
...
-                idleTimeout = idleTimeout,
-                requestTimeout = requestTimeout,
-                scheduler = scheduler,
-                ec = executionContext
-              ),
-              shutdown.guarantee(shutdownS)
-            )
-        }
-    }
+  def allocate(implicit F: ConcurrentEffect[F]): F[(Client[F], ExitCase[Throwable] => F[Unit])] =
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClientBuilder.scala]: I think I saw discussion of this return type fly by on Gitter recently, but I can't find it now.  Do we want the `ExitCase` here or nah?

**Comment by rossabaker** [comment]
Build failures are #2191, #2192.

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClientBuilder.scala (hunk 2 above)
Comment (ChristopherDavenport) [reply blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClientBuilder.scala]: Do we care about ExitCase ever?

File: core/src/main/scala/org/http4s/internal/package.scala [hunk 3]
```diff
@@ -117,4 +119,14 @@ package object internal {
           }
       }
     }
+
+  private[http4s] def allocated[F[_], A](resource: Resource[F, A])(
```
Comment (ChristopherDavenport) [inline core/src/main/scala/org/http4s/internal/package.scala]: This is too useful, whats the catch? The stronger constraint?

Comment (rossabaker) [reply core/src/main/scala/org/http4s/internal/package.scala]: That's the only downside I can see.

**Pull request merged**